checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -134,229 +413,786 @@ protected boolean checkOverride(MethodTree overriderTree,
-                reportFailure("override.holding.invalid.holdingonentry", overriderTree, enclosingType, overridden, overriddenType, null, null);
-            } else if (!overriddenHoldingOnEntryLocks.containsAll(overriderHoldingOnEntryLocks)) {
-                isValid = false;
-                reportFailure("override.holding.invalid", overriderTree, enclosingType, overridden, overriddenType, overriderHoldingOnEntryLocks, overriddenHoldingOnEntryLocks);
+
+
+    /**
+     * Checks that the field or array access is legal by checking that the locks
+     * in the access's expression are held.
+     *
+     * @param accessTree field or array access tree to check (may be an identifier tree of a field)
+     * @param treeToReportErrorAt Tree whose location is used to report the error
+     * @param expressionNode Node of the field or array access's expression
+     */
+    private void checkFieldOrArrayAccess(ExpressionTree accessTree, Tree treeToReportErrorAt, Node expressionNode) {
+        AnnotatedTypeMirror atmOfReceiver = atypeFactory.getReceiverType(accessTree);
+        if (treeToReportErrorAt != null && atmOfReceiver != null) {
+            AnnotationMirror gb = atmOfReceiver.getEffectiveAnnotationInHierarchy(GUARDEDBYUNKNOWN);
+            if (gb == null) {
+                ErrorReporter.errorAbort("LockVisitor.checkFieldOrArrayAccess: gb cannot be null");
             }
-        } else {
-            if (!overriderLocks.isEmpty()) {
-                if (!overriddenLocks.containsAll(overriderLocks)) {
-                    isValid = false;
-                    reportFailure("override.holding.invalid", overriderTree, enclosingType, overridden, overriddenType, overriderLocks, overriddenLocks);
-                }
-            } else if (!overriddenLocks.containsAll(overriderHoldingOnEntryLocks)) {
-                isValid = false;
-                reportFailure("override.holding.invalid", overriderTree, enclosingType, overridden, overriddenType, overriderHoldingOnEntryLocks, overriddenLocks);
+
+            if (AnnotationUtils.areSameByClass(gb, checkerGuardedByClass)) {
+                Set<Pair<String, String>> preconditions = generatePreconditionsBasedOnGuards(atmOfReceiver);
+                checkPreconditions(treeToReportErrorAt, expressionNode, preconditions);
+            } else if (AnnotationUtils.areSameByClass(gb, checkerGuardSatisfiedClass)){
+                // Can always dereference if type is @GuardSatisfied
+            } else {
+                // Can never dereference for any other types in the @GuardedBy hierarchy
+                String annotationName = gb.toString();
+                annotationName = annotationName.substring(annotationName.lastIndexOf('.') + 1 /* +1 to skip the last . as well */);
+                checker.report(Result.failure(
+                        "cannot.dereference",
+                        accessTree.toString(),
+                        "annotation @" + annotationName), accessTree);
```

Comment:
Thanks for catching this. I think at one point I had two places I issued a cannot.dereference error, and I wanted slightly different messages for each place. But even then I should have split it into two different error messages for internationalizaation. Fixed.